src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs

```diff
@@ -572,17 +563,11 @@ private static RuntimeTypeHandle GetDataContractAdapterTypeHandle(RuntimeTypeHan
                 if (type.IsInterface && !CollectionDataContract.IsCollectionInterface(type))
                     type = Globals.TypeOfObject;
 
-                lock (s_initBuiltInContractsLock)
+                return s_typeToBuiltInContract.GetOrAdd(type, static (Type key) =>
                 {
-                    s_typeToBuiltInContract ??= new Dictionary<Type, DataContract?>();
-
-                    if (!s_typeToBuiltInContract.TryGetValue(type, out DataContract? dataContract))
-                    {
-                        TryCreateBuiltInDataContract(type, out dataContract);
-                        s_typeToBuiltInContract.Add(type, dataContract);
-                    }
+                    TryCreateBuiltInDataContract(key, out DataContract? dataContract);
```

Comment:
TryCreateBuiltInDataContract seems thread safe (it only contains a switch statement and creates new DataContracts) so it should not be a problem to call it concurrently or the slim chance of creating duplicate instances